builtin/common/after.lua

```diff
@@ -9,42 +107,55 @@ core.register_globalstep(function(dtime)
 		return
 	end
 
-	time_next = math.huge
-
-	-- Iterate backwards so that we miss any new timers added by
-	-- a timer callback.
-	for i = #jobs, 1, -1 do
-		local job = jobs[i]
-		if time >= job.expire then
-			core.set_last_run_mod(job.mod_origin)
-			job.func(unpack(job.arg))
-			local jobs_l = #jobs
-			jobs[i] = jobs[jobs_l]
-			jobs[jobs_l] = nil
-		elseif job.expire < time_next then
-			time_next = job.expire
+	-- Remove the expired jobs.
+	local expired = remove_first_jobs()
+
+	-- Remove other expired jobs and append them to the list.
+	while true do
+		time_next = expiries[1] or math.huge
+		if time_next > time then
+			break
+		end
+		list_append_list(expired, remove_first_jobs())
+	end
+
+	-- Run the callbacks afterward to prevent infinite loops with core.after(0, ...).
+	local last_expired = expired.list_end
+	while true do
+		core.set_last_run_mod(expired.mod_origin)
+		expired.func(unpack(expired, 1, expired.n_args))
+		if expired == last_expired then
+			break
 		end
+		expired = expired.list_next
 	end
 end)
 
+local job_metatable = {__index = {}}
+
+local function dummy_func() end
+function job_metatable.__index:cancel()
```

Comment:
Maybe just remove the job from `job_map`?

Reply:
Why not use doubly linked list instead? It supports remove operation.

Reply:
Removal from the list is not needed for anything else. It's easier to just replace the function with a dummy.

Reply:
Easier for sure, but if you create a lot of jobs and cancel them later, you will still traverse the list and execute dummy functions. Actual removal should be more efficient in that case.

Reply:
There is also the special case where there is only one job in a list and it gets cancelled. Here the list would have to be removed from `job_map`. I don't think using a doubly linked list is worth it. I doubt jobs will be cancelled much anyway.

Reply:
I know the game that uses cancel a lot and suffers from it. But alright, your code - your rules :)

Reply:
Before this PR, each step where the `after` queue was checked was `O(number of total jobs)`. This includes cancelled jobs. After this PR, each step where a check occurs is, I believe, `O((number of different expiring job times) * log(number of different job times))`. This also includes cancelled jobs, but the better complexity should make them less costly.

Reply:
Yeah, it already reduces the overhead from canceled jobs a lot. But I feel it could be better.